### PR TITLE
Fix epw holidays header processing and fix eio reporting for holidays

### LINKED ISSUE #9743: epw file with holidays sometimes terminates with a "** Severe ** bad allocation" error
Issue overview
--------------
I am running e+9.5 with exactly the same IDF, but sometimes I get errors.

I run e+ 9.5 on the command line as follows
```
> energyplus -x -w weather.epw Input.idf
```
Why does the error occur even if I run e+ under the same conditions?
I don't know the cause of the error at all, does anyone know what the cause is?

The files used for the calculations are attached.
[mydata.zip](https://github.com/NREL/EnergyPlus/files/10031399/mydata.zip)

capture

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #9743 
 - Fixes error in epw holidays header processing introduced in v9.4 by #8157 (wrong placement of variable declaration). The defect file now runs to completion in debug and release.
 - Also fixes eio output for Special Days - should be one line of output per day.
 - Add DST and holidays to MadeUpLeapYear.epw header.
 - Turn on weather file DST and holidays in _5ZoneAirCooled_LeapYear_annual.idf integration test.
 - Merge various lists of day types into a single enum and associated name arrays.
 - Various other code modernization.

Diffs
-------------
First commit adding holidays to leap year test epw should cause _5ZoneAirCooled_LeapYear_annual to fail.
https://github.com/NREL/EnergyPlus/commit/f5e8a6bd900205a185920ad325cee273c64d9a9f#comments
image
and the eio special days output is garbage
```
+Environment:Special Days,CHRISTMAS DAY,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:Special Days,,,InputFile,13/-366,  0Environment:WarmupDays,  6
```

Now the special days are listed in separate lines as intended.
```
+Environment:Daylight Saving,Yes,WeatherFile,03/14,11/07
+Environment:Special Days,NEW YEAR'S DAY,Sunday,WeatherFile,01/01,  1
+Environment:Special Days,MLK JR BIRTHDAY,Sunday,WeatherFile,01/15,  1
+Environment:Special Days,PRESIDENTS' DAY,Sunday,WeatherFile,02/15,  1
+Environment:Special Days,MEMORIAL DAY,Sunday,WeatherFile,05/31,  1
+Environment:Special Days,JUNETEENTH,Sunday,WeatherFile,06/20,  1
+Environment:Special Days,INDEPENDENCE DAY,Sunday,WeatherFile,07/04,  1
+Environment:Special Days,LABOR DAY,Sunday,WeatherFile,09/06,  1
+Environment:Special Days,COLUMBUS DAY,Sunday,WeatherFile,10/11,  1
+Environment:Special Days,VETERANS DAY,Sunday,WeatherFile,11/11,  1
+Environment:Special Days,THANKSGIVING DAY,Sunday,WeatherFile,11/25,  1
+Environment:Special Days,CHRISTMAS DAY,Sunday,WeatherFile,12/26,  1
```

Other eio diffs are from certain design day day type names changing case. Previously, for this report, the day type names were taken from 2 separate string arrays, one for regular weekdays which was all uppercase, and the special day types were taken from a different string array which was not uppercase. Now the code draws from a single string array for all day types. 
```
-Environment,CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB,SizingPeriod:DesignDay,01/21,01/21,MONDAY,1,N/A,N/A,N/A,N/A,N/A,N/A,Clark and Allen
+Environment,CHICAGO_IL_USA ANNUAL HEATING 99% DESIGN CONDITIONS DB,SizingPeriod:DesignDay,01/21,01/21,Monday,1,N/A,N/A,N/A,N/A,N/A,N/A,Clark and Allen

```

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/WeatherManager.cc [hunk 1]
```diff
@@ -8602,497 +8488,522 @@ namespace WeatherManager {
...
+            // A2, \field Daylight Saving Start Day
+            // A3, \field Daylight Saving End Day
+            // N1, \field Number of Holidays
+            // A4, \field Holiday 1 Name
+            // A5, \field Holiday 1 Day
+            // etc.
+            // Start with Minimum number of NumHdArgs
+            uppercase(Line);
+            NumHdArgs = 4;
+            int CurCount = 0;
```
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:8696]: Here's the one-line fix for the defect. When declarations were localized, this one was put too far down inside the special day loop so it was always zero.

File: src/EnergyPlus/WeatherManager.cc [hunk 2]
```diff
@@ -8602,497 +8488,522 @@ namespace WeatherManager {
...
+                        ShowContinueError(state, format("...invalid header={}", epwHeaders[static_cast<int>(headerType)]));
+                        ShowContinueError(state, "...Setting Weather File DST to false.");
+                        state.dataWeatherManager->EPWDaylightSaving = false;
                     }
 
-                    int PMonth;
-                    int PDay;
-                    int PWeekDay;
-                    bool IOStatus;
-                    int CurCount = 0;
```
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:8807]: This was the wrong spot.

File: src/EnergyPlus/WeatherManager.hh [hunk 3]
```diff
@@ -71,6 +72,20 @@ struct EnergyPlusData;
 
 namespace WeatherManager {
 
+    enum class EpwHeaderType
```
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.hh:75]: New enum class for epwheaders.

File: src/EnergyPlus/WeatherManager.cc [hunk 4]
```diff
@@ -96,35 +96,38 @@ namespace WeatherManager {
...
     // This module contains all of the weather handling routines for
     // EnergyPlus.  That includes getting user input, defining design day
     // weather, retrieving data from weather files, and supplying the
     // outdoor environment for each time step.
 
-    // METHODOLOGY EMPLOYED:
-    // Setting up the design days is similar to BLAST/IBLAST.  Reading the
-    // BLAST weather files is similar to that code in BLAST/IBLAST.  The EnergyPlus
-    // Weather file (EPW) is new code.
+    constexpr std::array<std::string_view, static_cast<int>(EpwHeaderType::Num)> epwHeaders({"LOCATION",
```
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:106]: Moved this here and made it a std::array, which led to some squirrel chases.

File: src/EnergyPlus/WeatherManager.cc [hunk 5]
```diff
@@ -96,35 +96,38 @@ namespace WeatherManager {
...
-    // and internal Evolutionary Engineering documentation.
+    static constexpr std::array<std::string_view, static_cast<int>(WaterMainsTempCalcMethod::Num)> waterMainsCalcMethodNames{
+        "Schedule", "Correlation", "CorrelationFromWeatherFile", "FixedDefault"};
 
-    // Data
+    static constexpr std::array<std::string_view, static_cast<int>(WaterMainsTempCalcMethod::Num)> waterMainsCalcMethodNamesUC{
+        "SCHEDULE", "CORRELATION", "CORRELATIONFROMWEATHERFILE", "FIXEDDEFAULT"};
 
-    Array1D_string const DaysOfWeek(7, {"SUNDAY", "MONDAY", "TUESDAY", "WEDNESDAY", "THURSDAY", "FRIDAY", "SATURDAY"}); // NOLINT(cert-err58-cpp)
-    std::map<std::string, WeekDay> weekDayLookUp{{"SUNDAY", WeekDay::Sunday},                                           // NOLINT(cert-err58-cpp)
```
Comment (mjwitte) [inline src/EnergyPlus/WeatherManager.cc:119]: Lots of daytype squirrels, and other enum/string sets.

File: src/EnergyPlus/ScheduleManager.hh [hunk 6]
```diff
@@ -68,7 +68,28 @@ struct EnergyPlusData;
 
 namespace ScheduleManager {
 
-    int constexpr MaxDayTypes(12);
+    enum class DayType
```
Comment (mjwitte) [inline src/EnergyPlus/ScheduleManager.hh:71]: New DayType enum in ScheduleManager that includes all of the standard day types (replaces one that was in WeatherManager that only had Sunday-Saturday. Has a dummy slot at zero to keep the daytype integers the same as before.

File: src/EnergyPlus/ScheduleManager.cc [hunk 7]
```diff
@@ -84,51 +84,61 @@ namespace ScheduleManager {
...
-                                               "Velocity",
-                                               "Capacity",
-                                               "Power",
-                                               "Availability",
-                                               "Percent",
-                                               "Control",
-                                               "Mode"});
+    // Day types are 1-based for EMS and output and other uses, so add a dummy
+    constexpr std::array<std::string_view, static_cast<int>(DayType::Num)> dayTypeNames{"dummy",
+                                                                                        "Sunday",
```
Comment (mjwitte) [inline src/EnergyPlus/ScheduleManager.cc:100]: Single set of arrays for dayTypeNames and DayTypeNamesUC. Replaces a whole bunch of other string arrays scattered throughout the code.

**Review by mjwitte — COMMENTED** [review]
Walkthrough . . . 
The defect fix was moving a single line that reset a counter to zero.
Everything else is code modernization and a new unit test.
Some highlights.

File: src/EnergyPlus/OutputProcessor.cc [hunk 8]
```diff
@@ -2587,7 +2587,7 @@ namespace OutputProcessor {
             if (PrintTimeStamp) {
                 CurDayType = state.dataEnvrn->DayOfWeek;
                 if (state.dataEnvrn->HolidayIndex > 0) {
-                    CurDayType = 7 + state.dataEnvrn->HolidayIndex;
+                    CurDayType = state.dataEnvrn->HolidayIndex;
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.cc:2590]: Yep, sooo much nicer to not have to manage the magic 7 offset.  :+1:

File: src/EnergyPlus/OutputProcessor.hh [hunk 9]
```diff
@@ -136,39 +136,6 @@ namespace OutputProcessor {
 
     constexpr int N_WriteTimeStampFormatData(100);
 
-    constexpr const char *DayTypes(int &d)
```
Comment (Myoldmopar) [inline src/EnergyPlus/OutputProcessor.hh:139]: Yay!

File: src/EnergyPlus/RuntimeLanguageProcessor.cc [hunk 10]
```diff
@@ -205,8 +205,14 @@ void InitializeRuntimeLanguage(EnergyPlusData &state)
...
-    state.dataRuntimeLang->ErlVariable(state.dataRuntimeLangProcessor->HolidayVariableNum).Value =
-        SetErlValueNumber(double(state.dataEnvrn->HolidayIndex));
+    // Subtract 7 from HolidayIndex to maintain compatability for EMS where 1=Holiday,2=SummerDesignDay, 3=WinterDesignDay, 4=CustomDay1,
+    // 5=CustomDay2, but not <0
+    if (state.dataEnvrn->HolidayIndex == 0) {
+        state.dataRuntimeLang->ErlVariable(state.dataRuntimeLangProcessor->HolidayVariableNum).Value = SetErlValueNumber(0.0);
+    } else {
+        state.dataRuntimeLang->ErlVariable(state.dataRuntimeLangProcessor->HolidayVariableNum).Value =
+            SetErlValueNumber(double(state.dataEnvrn->HolidayIndex - 7));
+    }
```
Comment (Myoldmopar) [inline src/EnergyPlus/RuntimeLanguageProcessor.cc:215]: Yeah yeah, EMS + Magic Numbers is coming down the pike  :+1:

File: src/EnergyPlus/ScheduleManager.hh (hunk 6 above)
Comment (Myoldmopar) [reply src/EnergyPlus/ScheduleManager.hh:71]: The Dummy workaround is fine for now, but maybe we can even eliminate that sometime.  Either way, this is a much better situation than before.

File: src/EnergyPlus/WeatherManager.hh [hunk 11]
```diff
@@ -291,38 +292,29 @@ namespace WeatherManager {
...
-        Array1D_int monWeekDay;
-        int numSimYears;              // Total Number of years of simulation to be performed
-        bool isLeapYear;              // True if Begin Year is leap year.
-        bool RollDayTypeOnRepeat;     // If repeating run period, increment day type on repeat.
-        bool TreatYearsAsConsecutive; // When year rolls over, increment year and recalculate Leap Year
-        bool actualWeather;           // true when using actual weather data
-        bool firstHrInterpUsingHr1;   // true for using Hour 1 for first hour interpolate; fals for using Hour 24
-
-        // Default Constructor
-        RunPeriodData()
```
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.hh:319]: :smile:

File: src/EnergyPlus/api/datatransfer.py [hunk 12]
```diff
@@ -1025,7 +1025,10 @@ def holiday_index(self, state: c_void_p) -> int:
         :param state: An active EnergyPlus "state" that is returned from a call to `api.state_manager.new_state()`.
         :return: An integer indicator for current day holiday type.
         """
-        return self.api.holidayIndex(state)
+        if self.api.holidayIndex(state) == 0:
+          return 0
```
Comment (Myoldmopar) [inline src/EnergyPlus/api/datatransfer.py]: I think I need to fix the indent here prior to merge.

File: src/EnergyPlus/WeatherManager.cc (hunk 4 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WeatherManager.cc:106]: Fun!

File: src/EnergyPlus/WeatherManager.cc [hunk 13]
```diff
@@ -5912,27 +5857,11 @@ namespace WeatherManager {
...
-                                                "TUESDAY",
-                                                "WEDNESDAY",
-                                                "THURSDAY",
-                                                "FRIDAY",
-                                                "SATURDAY",
-                                                "HOLIDAY",
-                                                "SUMMERDESIGNDAY",
-                                                "WINTERDESIGNDAY",
-                                                "CUSTOMDAY1",
-                                                "CUSTOMDAY2"});
```
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc:5934]: Sheesh.  Good stuff!

File: src/EnergyPlus/WeatherManager.cc [hunk 14]
```diff
@@ -6569,27 +6496,15 @@ namespace WeatherManager {
...
-        static std::map<DDHumIndType, std::string> const DDHumIndTypeStringRep{{DDHumIndType::WetBulb, "Wetbulb [C]"},
```
Comment (Myoldmopar) [inline src/EnergyPlus/WeatherManager.cc:6585]: Yay, another map gone!

File: src/EnergyPlus/WeatherManager.cc (hunk 1 above)
Comment (Myoldmopar) [reply src/EnergyPlus/WeatherManager.cc:8696]: One line fix plus a nice little cleanup in the code.  :+1:   Thanks @mjwitte !

**Review by Myoldmopar — APPROVED** [review]
This looks great!  It grew large enough that you can't go through the changes line by line, but it looks like you've covered everything I'd expect to see changed and more.  Local testing ongoing, and I think I need to tweak the Python changes, but it appears to be in really good shape.

**Comment by Myoldmopar** [comment]
Everything looks great locally, just one table diff from the IDF that changed.  I pushed up the python indentation, I'll let one or two CIs finish at least before merging.  Thanks @mjwitte

**Comment by Myoldmopar** [comment]
Sames diffs as before, this thing is good to go.  Thanks @mjwitte

**Pull request merged**